sdk/core/core/src/az_span.c

```diff
@@ -618,3 +618,88 @@ _az_span_scan_until(az_span self, _az_predicate predicate, int32_t* out_index)
   }
   return AZ_ERROR_ITEM_NOT_FOUND;
 }
+
+AZ_NODISCARD az_span _az_span_trim_white_space(az_span source)
+{
+  // az_span source with size 0 is allowed. For instance, a header value can be AZ_SPAN_NULL
+  if (az_span_size(source) == 0)
+  {
+    return source;
+  }
+
+  // remove white spaces from left
+  az_span left_trim = _az_span_trim_white_space_from_start(source);
+  // calculate the offset after moving after white spaces
+  int32_t offset = az_span_size(source) - az_span_size(left_trim);
+  // remove from right
+  az_span right_trim = _az_span_trim_white_space_from_end(source);
+
+  if (offset == 0)
+  {
+    // nothing removed from left, return
+    return left_trim;
+  }
+
+  // slice right with the offset
+  return az_span_slice_to_end(right_trim, offset);
+}
+
+AZ_NODISCARD AZ_INLINE bool _az_is_white_space(uint8_t c)
+{
+  switch (c)
+  {
+    case ' ':
+    case '\t':
+    case '\n':
+    case '\r':
+      return true;
+  }
+  return false;
+}
+
+AZ_NODISCARD az_span _az_span_trim_white_space_from_start(az_span source)
+{
+  // az_span source with size 0 is allowed. For instance, a header value can be AZ_SPAN_NULL
+  if (az_span_size(source) == 0)
+  {
+    return source;
+  }
+
+  // loop from start to the first non white space
+  uint8_t* source_ptr = az_span_ptr(source);
+  for (int32_t index = 0; index < az_span_size(source); ++index)
+  {
+    if (!_az_is_white_space(source_ptr[index]))
+    {
+      return az_span_slice_to_end(source, index);
+    }
+  }
+  // reaching here means it was all white spaces
+  // Return az_span of size 0, using AZ_SPAN_NULL for now so we dont call slice
+  return AZ_SPAN_NULL;
```

Comment:
I don't think this is a good idea. We should return the slice instead.

Reply:
slice would return az_span with valid ptr but size 0.
I talked about this with @JeffreyRichter  and agree that ptr is not useful, so we just need the size 0 span.
We document that we return az_span with size 0 if there were only white spaces.  No need to mention what the pr is

Reply:
Why don't we side step this decision by breaking, and just slicing at the end?
```diff
  // loop from start to the first non white space
  uint8_t* source_ptr = az_span_ptr(source);
+ int32_t index = 0;
+ for (; index < az_span_size(source); ++index)
- for (int32_t index = 0; index < az_span_size(source); ++index)
  {
    if (!_az_is_white_space(source_ptr[index]))
    {
-     return az_span_slice_to_end(source, index);
+     break;
    }
  }
-  // reaching here means it was all white spaces
-  // Return az_span of size 0, using AZ_SPAN_NULL for now so we dont call slice
-  return AZ_SPAN_NULL;
+ return az_span_slice_to_end(source, index);
``

Reply:
For `_az_span_trim_white_space_from_end`, using gcc, that results in 12 fewer instructions and then you don't need that extra comment.
https://godbolt.org/z/WPdoCb

https://www.diffchecker.com/qcDvNKul